python/ray/serve/tests/test_callback.py

```diff
@@ -216,23 +220,33 @@ def test_http_proxy_calllback_failures(ray_instance, capsys):
     """Test http proxy into unhealthy state when callback function fails"""
 
     try:
-        serve.start()
+        serve.start(detached=True)
     except RayActorError:
         # serve.start will fail because the http proxy is not started successfully
         # and client use proxy handle to check the proxy readiness, so it will raise
         # RayActorError.
         pass
 
-    # 1s sleep to be long enough for proxy unhealthy message logged.
-    time.sleep(1)
+    client = get_global_client()
+
+    def check_http_proxy_keep_restarting():
+        # The proxy will be under "STARTING" status and keep restarting.
+        prev_actor_id = None
+        while True:
+            serve_details = ServeInstanceDetails(**client.get_serve_details())
+            for _, proxy_info in serve_details.http_proxies.items():
+                if proxy_info.status != HTTPProxyStatus.STARTING:
+                    return False
+                if prev_actor_id is None:
+                    prev_actor_id = proxy_info.actor_id
+                    break
+                elif prev_actor_id != proxy_info.actor_id:
+                    return True
+
+    wait_for_condition(check_http_proxy_keep_restarting)
 
     captured = capsys.readouterr()
     assert "this is from raise_error_callback" in captured.err
-    # Error message printed when http proxy in STARTING state and not
-    # able to be started.
-    assert (
-        "Unexpected actor death when checking readiness of HTTP Proxy" in captured.err
```

Comment:
Why are we loosen the condition? Is this no longer hit?

Reply:
I try to avoid to use checking log msg in the unit test, it is not reliable. The purpose of this test is to make sure proxy fail to start and keep restarting, so i change another way to verify it.

Reply:
Non-blocker: Maybe we can add some note on the this saying that's the desired behavior? Maybe in the docstring of this test. It's not immediately obvious to me the proxy restarting has to do with callback is failing (not before I read the code more XP)

Reply:
Updated test description.

Reply:
Thanks Sihan!